File: 704_ez_binary_search.py
#%%
from typing import List
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        l = 0
        r = len(nums) - 1 
        while(r >= l):
            c = ( r + l) //2
            if (target < nums[c]):
                r = c - 1
                print(f"Right is {nums[r]}")
            elif (target > nums[c]):
                l = c + 1
                print(f"Left is {nums[l]}")
            else :
                return c
            c = (r + l) //2 
            print(f"Center becomes {nums[c]}")
            if (target > nums[r]) or (target < nums[l]):
                return -1
        return -1

nums = [-1,0,3,5,9,12]
target = 10
sol = Solution().search(nums, target)
print(sol)       
# %%

File: 704_ez_Binary_Search.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int search(vector<int>& nums, int target) {
        int l = 0;
        int r = nums.size() - 1;
        if (target < nums[l] || target >  nums[r]) return -1;
        while(l <= r){
            int mid =  l - (l - r) / 2;
            if (target == nums[mid]){
                return mid;
            }
            if (target < nums[mid]){
                r = mid - 1;
            }
            else{
                l = mid + 1;
            }
        }
        return -1;
    }
};

int main(){
    Solution ans;
    vector<int> nums = {-2,0,3,5};
    int target = 5;
    int gg = ans.search(nums, target);
    cout << gg;
}